### PR TITLE
Rework resolving references and namespace creation in modules

### PR DESCRIPTION
**Pull request by zherczeg**
- Properties of namespace objects are ordered in lexical order
- The namespace object is created and imports are bound during linking phase
- Namespace imports are changed to read-only
- Reduced memory consumption

### CODE REVIEW

File: jerry-core/ecma/base/ecma-module.c [hunk 1]
```diff
@@ -156,221 +247,290 @@ ecma_module_resolve_set_cleanup (ecma_module_resolve_set_t *set_p) /**< resolve
...
+ecma_module_resolve_import (ecma_module_resolve_result_t *resolve_result_p, /**< [in,out] resolve result */
+                            ecma_module_resolve_set_t *resolve_set_p, /**< resolve set */
+                            ecma_module_t *module_p, /**< base module */
+                            ecma_string_t *local_name_p) /**< local name */
+{
+  ecma_module_node_t *import_node_p = module_p->imports_p;
+
+  while (true)
+  {
+    JERRY_ASSERT (import_node_p != NULL);
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply jerry-core/ecma/base/ecma-module.c:350]: It should never happen. Every local export must be a local binding or an imported binding. The formers are checked earlier.

File: jerry-core/ecma/base/ecma-module.c [hunk 2]
```diff
@@ -156,221 +247,290 @@ ecma_module_resolve_set_cleanup (ecma_module_resolve_set_t *set_p) /**< resolve
...
+ */
+static bool
+ecma_module_resolve_import (ecma_module_resolve_result_t *resolve_result_p, /**< [in,out] resolve result */
+                            ecma_module_resolve_set_t *resolve_set_p, /**< resolve set */
+                            ecma_module_t *module_p, /**< base module */
+                            ecma_string_t *local_name_p) /**< local name */
+{
+  ecma_module_node_t *import_node_p = module_p->imports_p;
+
+  while (true)
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply jerry-core/ecma/base/ecma-module.c:348]: See above. This loop must find the import.

File: jerry-core/ecma/base/ecma-module.c [hunk 3]
```diff
@@ -991,6 +1257,22 @@ ecma_module_link (ecma_module_t *module_p, /**< root module */
...
+                                                              ECMA_OBJECT_TYPE_CLASS);
+
+      namespace_object_p->type_flags_refs &= (ecma_object_descriptor_t) ~ECMA_OBJECT_FLAG_EXTENSIBLE;
+
+      ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) namespace_object_p;
+      ext_object_p->u.cls.type = ECMA_OBJECT_CLASS_MODULE_NAMESPACE;
+      ECMA_SET_INTERNAL_VALUE_POINTER (ext_object_p->u.cls.u3.value, module_p);
+
+      current_module_p->namespace_object_p = namespace_object_p;
+      ecma_deref_object (namespace_object_p);
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply jerry-core/ecma/base/ecma-module.c:1273]: The namespace_object_p should be marked by the GC. See:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/ecma/base/ecma-gc.c#L907

**Comment by zherczeg** [comment]
Thank you for the review, patch is updated.

**Review by rerobika — APPROVED** [review]

**Review by galpeter — APPROVED** [review]
lgtm

**Pull request merged**